### PR TITLE
Fix/env disable mkldnn cache map

### PR DESCRIPTION
**Pull request by azai91**
## Description ##
MKLDNN by default caches mkldnn memory and primitives to performance reasons. The cache looks uses input shape as a key. However, when models accept variable size inputs, the cache constantly grows and is not cleared. A flag is added MXNET_MKLDNN_CACHE_ENABLED that will disable this.

In a future PR, we can clear the cache automatically if it growths a certain length.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by stu1130** [comment]
@mxnet-label-bot add [pr-awaiting-review]
Thanks for your contribution @azai91

File: docs/faq/env_var.md [hunk 1]
```diff
@@ -195,6 +195,10 @@ When USE_PROFILER is enabled in Makefile or CMake, the following environments ca
   - Values: 0, 1 ```(default=1)```
   - Flag to enable or disable MKLDNN accelerator. On by default.
   - Only applies to mxnet that has been compiled with MKLDNN (```pip install mxnet-mkl``` or built from source with ```USE_MKLDNN=1```)
+  
+* MXNET_MKLDNN_CACHE_SIZE
+  - Values: Int ```(default=-1)```
+  - Flag to set MKLDNN cache size. Default is -1 which means cache size is unbounded. Should only be set if your model has variable input shapes, as cache size may grow unbounded.
```
Comment (anirudh2290) [inline docs/faq/env_var.md]: can you address the comment from https://github.com/apache/incubator-mxnet/pull/12746

**Comment by TaoLv** [comment]
#12746 is deprecated?

**Comment by pengzhao-intel** [comment]
It's better to continuous work on #12746 which includes several very useful comments.

**Comment by anirudh2290** [comment]
@azai91 found some lint and CI specific issues and thus moved to a new PR and deprecated the old one. @azai91 can you comment more ?

**Comment by azai91** [comment]
I had to revert to an older commit however this causes issues with the ci since it was reusing a cached version. Created a new branch that would bypass this

**Comment by anirudh2290** [comment]
Hi @pengzhao-intel , @TaoLv @ZhennanQin . I think because of the issues that Alex mentioned it seems fine to move to this PR. The outstanding comments are with respect to approach and the final decision on the PR is to take the approach suggested: "Use the unordered_set that drops elements randomly. I have tested this and it does not cause regression." @azai91 has implemented that in this PR. Can you please take a look.

**Comment by anirudh2290** [comment]
Ran some performance numbers which I have put below:

MXNET_MKLDNN_CACHE_SIZE=72 python  --network resnet-152

INFO:root:batch size  1, dtype float32, images/sec: 7.667724
INFO:root:batch size 32, dtype float32, images/sec: 19.753404
INFO:root:batch size 64, dtype float32, images/sec: 19.294091
INFO:root:batch size 128, dtype float32, images/sec: 18.829649
INFO:root:batch size 256, dtype float32, images/sec: 20.142726


INFO:root:batch size  1, dtype float32, images/sec: 7.703478
INFO:root:batch size 32, dtype float32, images/sec: 19.792880
INFO:root:batch size 64, dtype float32, images/sec: 20.017551
INFO:root:batch size 128, dtype float32, images/sec: 19.261422
INFO:root:batch size 256, dtype float32, images/sec: 20.087439

Latency
MXNET_MKLDNN_CACHE_SIZE=0 python example/image-classification/benchmark_score.py --network resnet-152 --batch-size 1
INFO:root:batch size  1, dtype float32, Time: 3.211219

MXNET_MKLDNN_CACHE_SIZE=-1 python example/image-classification/benchmark_score.py --network resnet-152 --batch-size 1
INFO:root:batch size  1, dtype float32, Time: 1.297572

MXNET_MKLDNN_CACHE_SIZE=44 python example/image-classification/benchmark_score.py --network resnet-152 --batch-size 1
INFO:root:batch size  1, dtype float32, images/sec: 1.332359

Note that for CACHE_SIZE 0 it triggers the cache removal and addition for every AddToCache call. The latency increases by 1.5 times in this case. Most customers wont set the cache size to so low.
Also, customers will be aware of the memory vs speed tradeoff when they use this env variable , so I think this should be fine.  For throughput we dont see a big difference.

**Comment by pengzhao-intel** [comment]
@azai91 @anirudh2290 thanks for the explanations :)

**Comment by TaoLv** [comment]
LGTM. Let's wait to see the CI results.

File: docs/faq/env_var.md [hunk 2]
```diff
@@ -195,6 +195,10 @@ When USE_PROFILER is enabled in Makefile or CMake, the following environments ca
   - Values: 0, 1 ```(default=1)```
   - Flag to enable or disable MKLDNN accelerator. On by default.
   - Only applies to mxnet that has been compiled with MKLDNN (```pip install mxnet-mkl``` or built from source with ```USE_MKLDNN=1```)
+  
+* MXNET_MKLDNN_CACHE_SIZE
+  - Values: Int ```(default=-1)```
+  - Flag to set MKLDNN cache size. Default is -1 which means cache size is unbounded. Should only be set if your model has variable input shapes, as cache size may grow unbounded. The number represents the number of items in the cache and is proportional to the number of layers that use MKLDNN and different input shape.
```
[the remark this answers is not in the document]
Comment (anirudh2290) [reply docs/faq/env_var.md]: agree . this confused me too on whether it was in bytes or num elements although size is used for num elements too (for eg. vector size in C++ denotes num of elements). WDYT @azai91 ?

[the remark this answers is not in the document]
Comment (azai91) [reply docs/faq/env_var.md]: alright, will rename to MXNET_MKLDNN_CACHE_NUM.

File: src/common/utils.h [hunk 3]
```diff
@@ -474,6 +474,10 @@ inline void LogStorageFallback(const nnvm::NodeAttrs& attrs,
 #if MXNET_USE_MKLDNN == 1
   if (!MKLDNNEnvSet()) common::LogOnce("MXNET_MKLDNN_ENABLED flag is off. "
                                        "You can re-enable by setting MXNET_MKLDNN_ENABLED=1");
+  if (GetMKLDNNCacheSize() != -1) common::LogOnce("MXNET_MKLDNN_CACHE_SIZE is set."
```
Comment (anirudh2290) [inline src/common/utils.h]: we need to change it here too.

**Pull request merged**